### PR TITLE
Handle WebSockets Across App & Refactor Frontend State

### PR DESCRIPTION
**Pull request by yimothysu**
Currently, a WebSocket connection is established in the `Terminal` component. This is problematic because WebSocket messages will be used to update the frontend for all types of messages (including chat, code editor, browser, and planner), not just the terminal as per https://github.com/OpenDevin/OpenDevin/issues/44.

This PR moves the WebSocket connection outside of the `Terminal` component so that all WebSocket messages can be propagated appropriately. To handle state changes from WebSocket messages, component state is moved into a Redux store. This also makes component files leaner and achieves better separation of the data layer from the view layer.

**TLDR: this PR adds Redux and refactors the frontend so that all components can make use of a shared WebSocket.**

### CODE REVIEW

**Comment by suryavirkapur** [comment]
I suggest we avoid using Redux for state management as it introduces unnecessary complexity. Instead, we should opt for a simpler and more case-appropriate solution like Jotai.

**Comment by yimothysu** [comment]
@suryavirkapur Can you give an example of unnecessary complexity in this commit?

I chose Redux for several reasons:
1. **Standardization:** Redux is the de facto standard for state management. Most frontend developers are familiar with Redux. There are numerous competing frameworks like MobX, Jotai, Recoil, etc. but it's most productive to use something everyone already knows.
2. **Simplicity**: Redux has been streamlined by the development of Redux Toolkit. Redux used to involve a lot of boilerplate in the past, but Redux Toolkit has eliminated that boilerplate.
3. **Ecosystem**: Redux enjoys a large and mature ecosystem. This includes development tools (e.g., time travel debugger), other libraries that extend its functionality, and overall community support.
4. **Scalability**: This project will grow more complex over time as we add logging, error handling, more complex WebSocket communication and API calls, etc. Redux Middleware allows us to easily extend the store to handle this, and the design of Redux allows us to easily reason about complex state management scenarios.

I prefer not using Jotai because a lot of its usage patterns encourage tightly coupling state to components, which makes test cases more cumbersome to write, involving many complex mocks and fixtures.

I want to avoid this becoming a framework debate though because it's unproductive. I've used Redux for a similar project and it worked well for the reasons above. Overall I think we should move forward unless there are major reservations.

**Comment by suryavirkapur** [comment]
@yimothysu Your argument is sound. I feel that Redux and Redux Toolkit might be overkill for this project. I suggest exploring [this implementation](https://github.com/suryavirkapur/OpenDevin/tree/main/frontend), which is more straightforward and analogous to using `useState` in React. Moreover, Jotai also boasts a well-established ecosystem and offers a simpler approach.

**Review by rbren — APPROVED** [review]
This lgtm, glad we're making the websockets more generic.

I don't have strong opinions on redux, but it does seem to be the "tried and true" library for state management in react.

**Comment by rbren** [comment]
This seems like a strong improvement over the current state, so let's get it in. Worth having the discussion re: Redux, maybe in the #frontend channel on Slack.
